jgrapht-core/src/main/java/org/jgrapht/alg/KDisjointShortestPaths.java

```diff
@@ -0,0 +1,328 @@
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.WeightedGraph;
+import org.jgrapht.graph.AsWeightedDirectedGraph;
+import org.jgrapht.graph.GraphWalk;
+
+/**
+ * The algorithm determines the k <em>disjoint</em> shortest simple paths in increasing order of
+ * weight. Weights can be negative (but no negative cycle is allowed). Only directed simple graphs
+ * are allowed. 
+ * <br>
+ * An undirected graph can be transformed to a directed simple graph using the following
+ * transformation:
+ * <br>
+ * For each pair of vertices V, U and an undirected edge {@literal E:V<-->U} with weight W:
+ * <ol>
+ * <li> Create vertices V_in, V_out, U_in and U_out
+ * <li> Create edges:
+ *     <ol>
+ *     <li> {@literal V_out-->U_in}, weight = W;
+ *  <li> {@literal U_out-->V_in}, weight = W;
+ *  <li> {@literal V_in-->V_out}, weight = 0;
+ *  <li> {@literal U_in-->U_out}, weight = 0;
+ *  </ol>
+ * </ol>
+ * <p>
+ * The algorithm is based on the Suurballe (later extended by Bhandari) algorithm, so is to find an
+ * Edge-disjoint shortest paths. In order to find a Vertex-disjoint shortest paths use the
+ * following transformation:
+ * <br>
+ * For each vertex V and a set of incoming edges E_in(V) and outgoing edges E_out(V):
+ * <ol>
+ * <li> Create vertex V_in
+ * <li> For every edge {@literal e:u-->V}, weight=W in E_in(V), create an edge {@literal e_in:u-->V_in}, weight=W
+ * <li> Create vertex V_out
+ * <li> For every edge {@literal e:V-->u}, weight=W in E_out(V), create an edge {@literal e_out:V_out-->u}, weight=W
+ * <li> Create an edge {@literal e_internal:V_in-->V_out} with weight=0.
+ * </ol>
+ *
+ * <p>
+ * The algorithm is running k sequential Bellman-Ford iterations to find the shortest path at each step.
+ * Hence, yielding a complexity of k*O(Bellman-Ford). Also, note that the provided graph is to be modified
+ * (edges and weights) during the path computation.
+ * 
+ * <p>
+ * For further reference, see <a href="https://www.nas.ewi.tudelft.nl/people/Fernando/papers/Wiley.pdf">Farabi Iqbal, Fernando A. Kuipers - 
+ * "Disjoint Paths In Networks"</a> which was the main reference for the code of this class.
+ * 
+ * @see BellmanFordShortestPath
+ *
+ * @author Assaf Mizrachi
+ * @since September 11, 2016
+ * 
+ * @param <V> Vertex
+ * @param <E> Edge
+ */
+public class KDisjointShortestPaths<V, E> {
+    /**
+     * Graph on which shortest paths are searched.
+     */
+    private Graph<V, E> graph;
+
+    private List<List<E>> pathList;
+
+    private int nPaths;
+
+    private V startVertex;
+
+    /**
+     * Creates an object to calculate k disjoint shortest paths between the start
+     * vertex and others vertices.
+     *
+     * @param graph
+     *            graph on which shortest paths are searched. Note: graph will

```

Comment:
I agree. But, the problem is that, unlike MaxFlow, Johnson and others, both Suurballe and Bhandari are modifying the graph structure (add/remove edges), not only the edge weights; and, unlike with modifying weights, non of the GraphDelegators can be used hide the effect from the original graph. Can we escape modifying the original graph nevertheless? Please advise.

Reply:
StoerWagnerMinimumCut provides an example of an existing algorithm which uses a working graph copy, including add/remove.

Reply:
Sure we can use a working graph copy but I thought we should leave this decision to the user. I will create a copy.

Reply:
Johnson's algorithm also adds a new vertex and a set of new edges. See the implementation to see how this can be done without affecting the original graph.